refactor(transformer): do not use `&mut` when `&` ref will do

overlookmotel opened the pull request:
#9253 enabled clippy's `needless_pass_by_ref_mut` rule, but kept it disabled for the `oxc_transformer` crate.

This rule is really good as using `&` instead of `&mut` may allow the compiler to make more optimizations in some cases, as `&` guarantees that value is not changed during the function call which takes `&`.

Enable this lint and fix all the warnings it highlights.

The diff is large, but it's purely replacing `&mut` with `&` in lots of places.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9266?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9266** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9266?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9266?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9265** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9264** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9264?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9266).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@Dunqing Just to flag that this PR touches many files in the transformer. Please can you merge this before any other transformer PRs, to avoid merge conflicts?

Boshen commented:
### Merge activity

* **Feb 21, 7:46 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(transformer): do not use `&mut` when `&` ref will do (#9266)

#9253 enabled clippy's `needless_pass_by_ref_mut` rule, but kept it disabled for the `oxc_transformer` crate.

This rule is really good as using `&` instead of `&mut` may allow the compiler to make more optimizations in some cases, as `&` guarantees that value is not changed during the function call which takes `&`.

Enable this lint and fix all the warnings it highlights.

The diff is large, but it's purely replacing `&mut` with `&` in lots of places.
